janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java

```diff
@@ -0,0 +1,122 @@
+public class JanusGraphMixedIndexCountStrategy extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy>
+    implements TraversalStrategy.ProviderOptimizationStrategy {
+    private static final JanusGraphMixedIndexCountStrategy INSTANCE = new JanusGraphMixedIndexCountStrategy();
+    private static final Set<Class<? extends ProviderOptimizationStrategy>> PRIORS = Collections.singleton(JanusGraphStepStrategy.class);
+
+    private JanusGraphMixedIndexCountStrategy() {
+    }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        if (TraversalHelper.onGraphComputer(traversal))
+            return;
+
+        TraversalHelper.getStepsOfClass(JanusGraphStep.class, traversal).forEach(originalGraphStep -> {
+            buildMixedIndexCountStep(originalGraphStep, traversal);
+        });
+    }
+
+    /**
+     * Check if a mixed index count step can be built, and if so, apply it.
+     *
+     * @param originalGraphStep
+     * @param traversal
+     */
+    private void buildMixedIndexCountStep(final JanusGraphStep originalGraphStep, final Traversal.Admin<?, ?> traversal) {
+        if (!hasCountGlobalStep(originalGraphStep)) {
+            return;
+        }
+        final JanusGraphMixedIndexCountStep<?> directQueryCountStep = new JanusGraphMixedIndexCountStep<>(traversal);
+
+        final JanusGraphTransaction tx = JanusGraphTraversalUtil.getTx(traversal);
+        // try to find a suitable mixed index and build a mixed index count query
+        MixedIndexCountQuery mixedIndexCountQuery = directQueryCountStep.buildGlobalGraphIndexQuery(originalGraphStep, tx);
+        if (mixedIndexCountQuery != null) {
+            applyMixedIndexCountStep(originalGraphStep, directQueryCountStep, traversal);
+        }
+    }
+
+    private boolean isEligibleToSkip(final Step currentStep) {
+        return currentStep instanceof IdentityStep || currentStep instanceof NoOpBarrierStep || currentStep instanceof HasContainerHolder;
+    }
+
+    private boolean hasCountGlobalStep(final GraphStep originalGraphStep) {
+        Step<?, ?> currentStep = originalGraphStep.getNextStep();
+        while (isEligibleToSkip(currentStep)) {
+            currentStep = currentStep.getNextStep();
+        }
+        return currentStep instanceof CountGlobalStep;
+    }
+
+    /**
+     * Apply mixed index count step in the traversal and remove "has" steps that already folded in the mixedIndexCountStep
+     *
+     * @param originalGraphStep
+     * @param mixedIndexCountStep
+     * @param traversal
+     */
+    private void applyMixedIndexCountStep(final GraphStep originalGraphStep, final JanusGraphMixedIndexCountStep mixedIndexCountStep,
+                                          final Traversal.Admin<?, ?> traversal) {
+        Step<?, ?> currentStep = originalGraphStep.getNextStep();
+        while (true) {
+            if (currentStep instanceof HasContainerHolder || currentStep instanceof CountGlobalStep) {
```

Comment:
I stumbled upon this logic and was wondering if it is correct. For example, a query with `count().count()` (which by the way is completely bonkers) should always return 1, but would be reduced to one `JanusGraphMixedIndexCountStep` here.

Reply:
That is a great point! I don't even know `count().count()` is legal. I'll check if the result is still correct in this case.

Reply:
Should be legal because `count()` simply produces a single traverser with a `count` value. This traverser itself can also be counted even if that makes no sense.
In addition, I am not entirely sure if there can exist other `HasContainerHolder` steps between the `JanusGraphStep` and the `CountGlobalStep`. For example, `JanusGraphStep` itself is a `HasContainerHolder` and so may be other steps. Let me take something like this as an example: `g.V().V().count()`. Since both `V()` are represented by `JanusGraphStep`, both satisfy the condition for this strategy.

Reply:
I rewrote this method and added a few test cases. Let me know if you come up with other corner cases.